Refactor delay parameters

shehbazj opened the pull request:
Added hardcoded delay values that correspond to real NVMe I/O device.
Set NVMe Program and Read macros in FTL layer to 0. These are used to compute `req->expire_time` while queueing requests in the completion queue.

amytai commented on hw/block/femu/femu.c:
```diff
@@ -478,11 +478,9 @@ static uint16_t nvme_io_cmd(FemuCtrl *n, NvmeCmd *cmd, NvmeRequest *req, int com
     case NVME_CMD_READ:
     case NVME_CMD_WRITE:
         if (n->femu_mode == FEMU_BLACKBOX_MODE) {
-		femu_debug("%s():bb mode calling - nvme_rw\n", __func__);
```
You can keep these in since they're just for debugging. Was there a specific reason why these lines are removed?

shehbazj replied:
this print statement was added by me in an earlier version. it is not part of FEMU's implementation. This becomes very verbose as its executed for each command while running benchmarks so I removed it.

amytai commented on hw/block/femu/femu.c:
```diff
@@ -1274,6 +1273,7 @@ static Property femu_props[] = {
     DEFINE_PROP_UINT16("did", FemuCtrl, did, 0x1f1f),
     DEFINE_PROP_UINT8("femu_mode", FemuCtrl, femu_mode, FEMU_DEF_NOSSD_MODE),
     DEFINE_PROP_UINT8("computation_mode", FemuCtrl, computation_mode, FEMU_DEF_NOCOMPUTATION_MODE),
+    DEFINE_PROP_UINT8("nscdelay_mode", FemuCtrl, nscdelay_mode, FEMU_DEF_NSCDELAY_MODE),
```
I recommend you add parameterizations for the read/write latencies as well, instead of hard-coding them with `#define` in `mem-backend.c` below.

shehbazj replied:
yes I will add read/write parameters. thank you.

amytai commented on hw/block/femu/ftl/ftl.h:
```diff
@@ -13,8 +13,10 @@ enum {
 
-    NAND_READ_LATENCY = 40000,
-    NAND_PROG_LATENCY = 200000,
+    NAND_READ_LATENCY = 0,
```
These should only be 0 if there all commands use the computational thread. I suggest you do not change these enums, and instead add another enum type for in-device computation latencies. For now, the in-device computation latency can be 0.

Then, in the code that adds latencies, you can determine which enum to add based on whether computation is on.

shehbazj replied:
these are valid points, I'll make a separate in-device latency enum. Thank you.

amytai commented on hw/block/femu/femu.c:
```diff
@@ -1274,6 +1273,7 @@ static Property femu_props[] = {
     DEFINE_PROP_UINT16("did", FemuCtrl, did, 0x1f1f),
     DEFINE_PROP_UINT8("femu_mode", FemuCtrl, femu_mode, FEMU_DEF_NOSSD_MODE),
     DEFINE_PROP_UINT8("computation_mode", FemuCtrl, computation_mode, FEMU_DEF_NOCOMPUTATION_MODE),
+    DEFINE_PROP_UINT8("nscdelay_mode", FemuCtrl, nscdelay_mode, FEMU_DEF_NSCDELAY_MODE),
```
How is `nscdelay_mode` different from `computation_mode`? Is there a case when `computation_mode=1` but `nscdelay_mode=0`?

shehbazj replied:
`nscdelay_mode` enables or disables the in device latency. You are correct in mentioning that `computation_mode=1 nscdelay_mode=0` is not useful. But the other three modes have usecases:

1. `computation_mode=0 nscdelay_mode=0` can be used to tweak PCIe latency (`NAND_READ_LATENCY` & `NAND_WRITE_LATENCY`) since device will be working at memory speed.  
2.  `computation_mode=0 nscdelay_mode=1` can be used to tweak on-disk latency and do host benchmarks. Finally
3. `computation_mode=1 nscdelay_mode=1` should be used for doing nsc benchmarks.

amytai replied:
I don't think `computation_mode=0 nscdelay_mode=1` is a useful mode, since `NAND_READ_LATENCY`, etc. capture the e2e latency (media latency + PCIe latency).

shehbazj replied:
Sorry for the confusion, here is my reasoning for the `computation_mode=0 nscdelay_mode=1` configuration - with `computation_mode=0` we run host. if we set `nscdelay_mode=0` while running host and then compare this with `computation_mode=1 nscdelay_mode=1` during `nsc`, we are adding media delay (`READ_LATENCY`, `WRITE_LATENCY`) during computation but not during host reads and writes (when `computation_mode=0`). This will lead to `host` accessing device faster than `nsc`, because in `host` mode it will read data from underlying disk at memory speed, but in `nsc` mode we will wait `READ_LATENCY` microseconds on each flash access. 

My understanding is that we need to split existing FEMU's `NAND_READ_LATENCY` into two - `PCIe_LATENCY` (which will be encountered by `host` on each block access but by `nsc` only once in the entire pointer chase operation as it sends the request to the device once and receives the pointer tail) and `FLASH_READ_LATENCY` (which will be encountered by `host` and `nsc` on each block access within the guest).

amytai replied:
We want to capture two modes:
1) host mode: `computation_mode=0`. This means every operation will be delayed `NAND_READ_LATENCY`, which is currently an e2e latency, i.e. PCIe latency + media latency
2) nsc mode: `computation_mode=1`. This means every read within the device will be delayed `READ_LATENCY`, but only incur 1 PCIe latency on return to host.

I think we agree on these two modes.

To support these two modes, you can take a variety of approaches:
a) Don't do anything for host mode: it takes the legacy FEMU path. For nsc mode, protect the `READ_LATENCY` overhead with a `if(computation_mode == 1)`. Then add a separate PCIe latency on return to host (btw to be precise, PCIe latency doesn't appear to be in this PR)
b) Introduce a new PCIe_latency overhead that is used in both host _and_ nsc mode. Eliminate `NAND_READ_LATENCY`, which is the legacy FEMU path. Replace this with PCIe_latency, and do NOT protect `READ_LATENCY` with an `if` statement. This way, the media latency (`READ_LATENCY`) will be added to _all_ operations, regardless if they are for nsc or host mode, and the PCIe latency will only be incurred on return-to-host.

I assume you're going with approach (a), because I do not see an additional PCIe latency. Furthermore, in either approach, I still don't see how you would use `nscdelay_mode`.

shehbazj replied:
this was an excellent suggestion, thank you!

shehbazj commented:
1. removed `nscdelay_mode` from the code and added `flash_read_delay` and `flash_write_delay` configurable parameters in mem-backend.c. `nscdelay_mode=0` is simulated by default when `flash_read_delay` and `flash_write_delay` are not set explicitly by the user.
2. Restored delay added by FEMU in the return path. and renamed it to PCIe delay.

amytai commented on hw/block/femu/mem-backend.c:
```diff
@@ -101,26 +105,32 @@ int femu_rw_mem_backend_bb(struct femu_mbe *mbe, QEMUSGList *qsg,
 	// address_space, current_address, buffer, length_of_dma_address, read_or_write_direction
 	if (!mbe->computation_mode) {
+		if (is_write) {
+			add_delay(flash_write_delay);
```
There's an opportunity here to refactor `add_delay` and `dma_memory_rw`. One way is to put the block:
```
if (is_write) {
			add_delay(flash_write_delay);
} else {
			add_delay(flash_read_delay);
		}
 if (dma_memory_rw(qsg->as, cur_addr, mb + mb_oft, cur_len, dir)) {
			error_report("FEMU: dma_memory_rw error");
 }
```

at the top of this logic (line 107). Do not protect it with `if (!mbe->computation_mode)`, because this media delay and the read/write will happen regardless of what mode we are in.
Then, you'll need to refactor your `do_pointer_chase` and `do_count_bits` such that they occur after the `dma`. This is cleaner because then the compute function will only be executing the "compute" portion, after the data has been read (expect for the pointer_chase case, where you will do additional dma inside the `do_pointer_chase` function).

shehbazj replied:
I see what you mean, I have refactored the code so that the first I/O takes place with delay irrespecive of computation mode and the subsiquent I/Os `add_delay` on a case by case basis. this makes sense to me. thank you.

shehbazj commented on hw/block/femu/ftl/ftl.h:
```diff
@@ -13,8 +13,8 @@ enum {
 
-    NAND_READ_LATENCY = 40000,
-    NAND_PROG_LATENCY = 200000,
+    PCIe_READ_LATENCY = 40000,
```
Yes this makes sense, I have added it to current commit. I will keep the different PCIe latency for larger workload while doing large I/Os later. thank you.

shehbazj commented on hw/block/femu/computation.c:
```diff
@@ -21,3 +21,9 @@ int get_disk_pointer(char *buf)
 	return p[0];
 }
 
+void inline add_delay(uint32_t micro_seconds) {
```
`add_delay` is now part of `mem-backend.c`. you are right, it is only used by `mem-backend.c`, I have made the function `static`.

shehbazj commented:
@amytai Added additional changes requested. Tested pointer chase with both `host` and `nsc` mode for loop based bugs. Please review, thank you!

amytai commented on hw/block/femu/mem-backend.c:
```diff
@@ -37,37 +37,44 @@ void femu_destroy_mem_backend(struct femu_mbe *mbe)
 	DMADirection dir = DMA_DIRECTION_FROM_DEVICE;
 
-	do
+	while (1)
```
`while (1)` seems dangerous. Instead, you can run the first step of the pointer chase outside of the while loop. Then put the while loop with your old condition (` while (c != 0 && c != END_BLOCK_MAGIC)`)

shehbazj replied:
refactored `while(1)` to `while (c != 0 && c != END_BLOCK_MAGIC)`.
Request your review. Thank you!

shehbazj commented on hw/block/femu/mem-backend.c:
```diff
@@ -99,28 +108,28 @@ int femu_rw_mem_backend_bb(struct femu_mbe *mbe, QEMUSGList *qsg,
+		error_report("FEMU: dma_memory_rw error");
+        }
+
+	// address_space, current_address, buffer, length_of_dma_address, read_or_write_direction
```
removed comment.
